web/src/main/webapp/lamp.ts

```diff
@@ -0,0 +1,412 @@
+    public static maxNumSamples = 100;
+
+    constructor(private selectedColumns: string[], private page: FullPage,
+                private schema: Schema, private remoteObject: RemoteTableObject) {
+        super("LAMP");
+        this.addTextField("numSamples", "No. control points", "Integer", "15");
+        this.addSelectField("controlPointSelection", "Control point selection", ["Random samples", "Category centroids"], "Random samples");
+        let catColumns = [""];
+        for (let i = 0; i < schema.length; i++)
+            if (schema[i].kind == "Category")
+                catColumns.push(schema[i].name);
+        this.addSelectField("category", "Category for centroids", catColumns, "");
+        this.addSelectField("controlPointProjection", "Control point projection", ["MDS"], "MDS");
+        this.setAction(() => this.execute());
+    }
+
+    private execute() {
+        let numSamples = this.getFieldValueAsInt("numSamples");
+        if (numSamples > LAMPDialog.maxNumSamples) {
+            this.page.reportError(`Too many samples. Use at most ${LAMPDialog.maxNumSamples}.`);
+            return;
+        }
+        let selection = this.getFieldValue("controlPointSelection");
+        let projection = this.getFieldValue("controlPointProjection");
+        let category = this.getFieldValue("category");
+
+        let rr: RpcRequest;
+        switch (selection) {
+            case "Random samples": {
+                rr = this.remoteObject.createSampledControlPointsRequest(numSamples, this.selectedColumns);
+                break;
+            }
+            case "Category centroids": {
+                rr = this.remoteObject.createCategoricalCentroidsControlPointsRequest(category, this.selectedColumns);
+                break;
+            }
+            default: {
+                this.page.reportError(`${selection} is not implemented.`);
+            }
+        }
+
+        let newPage = new FullPage();
+        this.page.insertAfterMe(newPage);
+
+        switch (projection) {
+            case "MDS": {
+                rr.invoke(new ControlPointsProjector(newPage, rr, this.remoteObject, this.selectedColumns));
+                break;
+            }
+            default: {
+                this.page.reportError(`${projection} is not implemented.`);
+            }
+        }
+    }
+}
+
+class ControlPointsProjector extends RemoteTableRenderer {
+    constructor(page, operation, private tableObject: RemoteTableObject, private selectedColumns) {
+        super(page, operation, "Sampling control points");
+    }
+
+    onCompleted() {
+        super.finished();
+        if (this.remoteObject == null)
+            return;
+        let rr = this.tableObject.createMDSProjectionRequest(this.remoteObject.remoteObjectId);
+        rr.invoke(new ControlPointsRenderer(this.page, rr, this.tableObject, this.remoteObject.remoteObjectId, this.selectedColumns));
+    }
+}
+
+class ControlPointsRenderer extends RpcReceiver<PointSet2D> {
```

Comment:
I think you should use Renderer here instead of RpcReceiver.

Reply:
It's now a Renderer (I'll push it soon), but I had to do some more verbose packaging of JSON objects to construct the RPC reply in the head node. This is because the result is not a result of a sketch/map, but is computed sequentially on the head node.

Maybe there should be a `Loop` type of computation, along with a `LoopResultObserver`, that receives/sends updates every n iterations of the loop. But, given my time constraints (and lack of experience with Rx) I don't think I can implement this.

Also, right now it's not really necessary, as the MDS is quite fast (as long as the number of samples is not too large).